typing/ctype.ml

```diff
@@ -3233,19 +3286,22 @@ let expand_head_trace env t =
 *)
 
 let filter_arrow env t l =
-  let t = expand_head_trace env t in
+  let t =
+    try expand_head_trace env t
+    with Unify_trace _ -> assert false
```

Comment:
This `assert false` is reachable:
```ocaml

module M = struct
  type t
end
type (_,_) eq = Refl: ('a,'a) eq
let f (x:M.t) (y: (M.t, int ->int) eq) =
  let Refl = y in
  if true then x else fun x -> x +1
```

Reply:
Thanks for the catch, I've rebased this bugfix into the first commit and added this test case.